File: strategy/cpp/example_1/strategy.cpp
#include <iostream>

// like an interface

class FlyBehavior 
{
public:
	virtual void fly() = 0;
};


class FlyWithWings : public FlyBehavior
{
public:
	virtual void fly() {
		std::cout << "I'm flying" << std::endl;
	}
};


class FlyNoWay : public FlyBehavior
{
public:
	virtual void fly() {
		std::cout << "I can't fly" << std::endl;
	}
};

class FlyRocketPowered : public FlyBehavior
{
public:
	virtual void fly() {
		std::cout << "I'm flying with a rocket" << std::endl;
	}
};

////////////////////////////////

class QuackBehavior
{
public:
	virtual void quack() = 0;
};

class Quack : public QuackBehavior
{
public:
	virtual void quack() {
		std::cout << "Quack" << std::endl;
	}
};

class MuteQuack : public QuackBehavior
{
public:
	virtual void quack() {
		std::cout << " Silence" << std::endl;
	}
};

////////////////////////////////
class Duck 
{
protected:
	FlyBehavior *flyBehavior;
	QuackBehavior *quackBehavior;
public:
	Duck() {}
	
	virtual void display() = 0;
	
	void performFly() {
		flyBehavior->fly();
	}

	void performQuack() {
		quackBehavior->quack();
	}

	void swim() {
		std::cout << "All ducks float" << std::endl;
	}

	void setFlyBehavior(FlyBehavior *fb) {
		flyBehavior = fb;
	}

	void setQuackBehavior(QuackBehavior *qb) {
		quackBehavior = qb;
	}
};

class MallardDuck : public Duck
{
public:
	MallardDuck() {
		quackBehavior = new Quack();
		flyBehavior = new FlyWithWings();
	}

	virtual void display() {
		std::cout << "I'm a real Mallard Duck" << std::endl;
	}
};

class ModelDuck : public Duck
{
public:
	ModelDuck() {
		flyBehavior = new FlyNoWay();
		quackBehavior = new Quack();
	}

	virtual void display() {
		std::cout << "I'm model duck" << std::endl;
	}
};

////////////////////////////////

int main() {
	Duck *mallard = new MallardDuck();
	mallard->performQuack();
	mallard->performFly();
	delete mallard;

	Duck *model = new ModelDuck();
	model->performFly();
	model->setFlyBehavior(new FlyRocketPowered());
	model->performFly();
	delete model;
	return 0;
}

File: strategy/python/example_1/strategy.py
# -*- coding: utf-8 -*-

#############################################

class Duck:
	_fly_behavior = None;
	_quack_behavior = None;

	def __init__(self):
		pass

	def display(self):
		pass

	def perform_fly(self):
		self._fly_behavior.fly()

	def perform_quack(self):
		self._quack_behavior.quack()

	def swim():
		print("All ducks float")

	@property
	def fly_behavior(self):
	    return self._fly_behavior
	@fly_behavior.setter
	def fly_behavior(self, value):
	    self._fly_behavior = value
	

#############################################

class FlyBehavior:

	def fly(self):
		pass


class FlyWithWings(FlyBehavior):
	
	def fly(self):
		print("I'm flying!")


class FlyNoWay(FlyBehavior):

	def fly(self):
		print("I can't fly")


class FlyRocketPowered(FlyBehavior):

	def fly(self):
		print("I'm flying with rocket")


#############################################

class QuackBehavior:

	def quack(self):
		pass


class Quack(QuackBehavior):

	def quack(self):
		print("Quack")


class MuteQuack(QuackBehavior):

	def quack(self):
		print("Silence")


class Squeak(QuackBehavior):

	def quack(self):
		print("Squeak")


#############################################


class MallardDuck(Duck):

	def __init__(self):
		self._quack_behavior = Quack()
		self._fly_behavior = FlyWithWings()


class ModelDuck(Duck):

	def __init__(self):
		self._fly_behavior = FlyNoWay()
		self._quack_behavior = Quack()

	def display(self):
		print("I'm a model duck")


#############################################

if __name__ == "__main__":
	mallard = MallardDuck()
	mallard.perform_quack()
	mallard.perform_fly()

	model = ModelDuck()
	model.perform_fly()
	model.fly_behavior = FlyRocketPowered()
	model.perform_fly()
